### PR TITLE
Add hexview on rz-diff

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This will add hexview on rz-diff

rz-diff-hexview

### CODE REVIEW

**Comment by XVilka** [comment]
1. I suggest to highlight the differences on both sides like `dhex` does it: 

image

2. A shortcuts and a helpscreen to quickly jump between
    - first diff in the file
    - last diff in the file
    - previous diff
    - next diff

**Comment by XVilka** [comment]
More notes:
- [ ] PgUp and PgDown should scroll for the entire screen/page
- [x] Scrolling order seems inverted: PgUp and ArrowUp scrolls down, while PgDown and ArrowDown scrolls up
- [x] It doesn't respect colors setting in `.rizinrc`
- [x] Probably use Unicode symbols if `scr.utf8` is set (and curvy edges when `scr.utf8.curvy` is set)
- [x] Pressing `?` in this mode should show the help screen, like in Visual mode of Rizin, with corresponding hotkeys

I also think it might be nice to have this feature in Rizin itself - probably between opened files (`o` output e.g.)

**Comment by XVilka** [comment]
- [x] We also should test how this behaves on Windows.

**Comment by wargio** [comment]
image

A few notes from windows:

- Tested colors and theme
- rzcons behave a bit weirdly on windows but is more than usable. all the features works

**Comment by wargio** [comment]
Extra notes: I can't distinguish PgUp and PgDown because the read char is the same as the Up and Down arrrows on both OS.

**Comment by XVilka** [comment]
Ok, now most of my points are fixed. Regarding the help and legend of the keys below: If `scr.utf8` is set, maybe use the arrow Unicode characters instead of the `/\`, `\/`, `<`, `>`? 
- https://github.com/rizinorg/rizin/blob/dev/librz/core/disasm.c#L36
- https://unicode-table.com/en/sets/arrow-symbols/

Also it didn't load the color theme installed in `~/.local`:
```
[i] ℤ rizin -H                                                                                                                                                                                                                    12:38:19 
RZ_VERSION=0.3.0-git
RZ_PREFIX=/home/akochkov/.local
RZ_MAGICPATH=/home/akochkov/.local/share/rizin/0.3.0-git/magic
RZ_INCDIR=/home/akochkov/.local/include/librz
RZ_LIBDIR=/home/akochkov/.local/lib64
RZ_LIBEXT=so
RZ_RCONFIGHOME=/home/akochkov/.config/rizin
RZ_RDATAHOME=/home/akochkov/.local/share/rizin
RZ_RCACHEHOME=/home/akochkov/.cache/rizin
RZ_LIBR_PLUGINS=/home/akochkov/.local/lib64/rizin/0.3.0-git
RZ_USER_PLUGINS=/home/akochkov/.local/share/rizin/plugins
RZ_USER_ZIGNS=/home/akochkov/.local/share/rizin/zigns
RZ_IS_PORTABLE=0
```

Also there is one miniscule leak if you build it with ASAN:
```
=================================================================
==195603==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 2 byte(s) in 1 object(s) allocated from:
    #0 0x7f2f4b89f967 in strdup (/lib64/libasan.so.6+0x59967)
    #1 0x7f2f4b7958b9 in visual_prompt ../librz/main/rz-diff.c:1750
    #2 0x7f2f4b7958b9 in prompt_offset_and_seek ../librz/main/rz-diff.c:1756
    #3 0x7f2f4b7958b9 in rz_diff_hex_visual ../librz/main/rz-diff.c:1909
    #4 0x7f2f4b7958b9 in rz_main_rz_diff ../librz/main/rz-diff.c:2058
    #5 0x7f2f4b557b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)

SUMMARY: AddressSanitizer: 2 byte(s) leaked in 1 allocation(s).
```

**Comment by XVilka** [comment]
Minor compiler warning:
```
librz/main/librz_main.so.0.3.0-git.p/rz-diff.c.o -c ../librz/main/rz-diff.c
../librz/main/rz-diff.c: In function ‘rz_main_rz_diff’:
../librz/main/rz-diff.c:2077:1: error: label ‘rz_main_rz_diff_end’ defined but not used [-Werror=unused-label]
 2077 | rz_main_rz_diff_end:
      | ^~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

File: librz/main/rz-diff.c [hunk 1]
```diff
@@ -183,6 +218,11 @@ static void rz_diff_show_help(bool usage_only) {
 		"              sections   | compares sections found in the files\n"
 		"              strings    | compares strings found in the files\n"
 		"              symbols    | compares symbols found in the files\n"
+		"  palette colors can be changed by adding the following lines\n"
```
Comment (XVilka) [inline librz/main/rz-diff.c:223]: Could you set them in the default theme at least? And maybe move this piece of the documentation in the `rz-diff` man page? It doesn't really belong in the `-h` output, I think.

File: librz/main/rz-diff.c [hunk 2]
```diff
@@ -1339,6 +1428,749 @@ static bool rz_diff_graphs_files(DiffContext *ctx) {
...
+
+static void prompt_offset_and_seek(DiffHexView *hview, ut64 minseek) {
+	char *value = visual_prompt(hview, " you can input an absolute offset or a relative offset by adding the prefix + or -\n offset");
+	if (value) {
+		const char *p = rz_str_trim_head_ro(value);
+		if (!IS_DIGIT(*p) && *p != '-' && *p != '+') {
+			free(value);
+			return;
+		}
+		st64 number = strtoll((*p == '+') ? p + 1 : p, NULL, 0);
```
Comment (XVilka) [inline librz/main/rz-diff.c:1777]: Maybe better to use `rz_num` here?

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
Please use the inverted background for differences if no colors are available.

File: librz/main/rz-diff.c (hunk 1 above)
Comment (wargio) [reply librz/main/rz-diff.c:223]: they are already in the default theme.
The only thing that is missing is updating the book.

File: librz/main/rz-diff.c (hunk 2 above)
Comment (wargio) [reply librz/main/rz-diff.c:1777]: no, because it's too messy and also calls the same function

**Comment by wargio** [comment]
> Please use the inverted background for differences if no colors are available.

done

**Comment by thestr4ng3r** [comment]
> Extra notes: I can't distinguish PgUp and PgDown because the read char is the same as the Up and Down arrrows on both OS.

How do visual mode and other applications handle this? Vim handles it fine for example.

**Comment by thestr4ng3r** [comment]
Resizing the terminal seems broken (works in rizin V):
Bildschirmfoto 2021-05-20 um 20 31 41

**Comment by wargio** [comment]
> Resizing the terminal seems broken (works in rizin V):
> Bildschirmfoto 2021-05-20 um 20 31 41

even when you press something?

**Comment by thestr4ng3r** [comment]
> even when you press something?

No, as soon as it refreshes once it's fixed. But if possible it should automatically do that on resize.

**Comment by wargio** [comment]
> > even when you press something?
> 
> No, as soon as it refreshes once it's fixed. But if possible it should automatically do that on resize.

git pull and try again

**Comment by thestr4ng3r** [comment]
> > > even when you press something?
> > 
> > 
> > No, as soon as it refreshes once it's fixed. But if possible it should automatically do that on resize.
> 
> git pull and try again

works :-)

Some other ideas/notes:
- [x] on the bottom right you can always see the cursor (or at least some white square), and scrolling quickly with a high key repeat sometimes flickers a bit with a visible cursor, maybe this is related
- [ ] hexdump width sometimes gets inconveniently large on a maximized window, maybe introduce `[` and `]` like in visual hexdump for resizing and another key pair for switching between powers of 2. Not sure how to best combine with the auto-resizing though.
- [x] It is missing the top row that indicates what column corresponds to which local byte offset


https://user-images.githubusercontent.com/1460997/119033387-721e4800-b9ad-11eb-9f22-4816117a3e6c.mov

**Comment by XVilka** [comment]
Looks much better. A couple more bugs appeared:
- Since you hide the cursor, please return it as it was when you exit `rz-diff` - it didn't show in the console if it was supposed to be shown even after the exit.
- It's too narrow on my screen by default now:

image

And one more suggestion - please add `vi`-like hotkeys for going to the beginning of the file and the end of it with `g`/`G`. I am not sure what current `G - Reset` hotkey means. 

Overall, it looks awesome now. Amazing job! :+1:

**Comment by XVilka** [comment]
Hmm, it now shows the cursor after quitting, but changed it's style. Before I had a block cursor, now I have a thin line after `rz-diff`.
Is it possible to just preserve the state of the cursor and restore as it was? If there is no such API in RzCons - maybe it worth to introduce one.

**Review by XVilka — APPROVED** [review]
LGTM as is. I suggest to merge this, and extend/improve in separate PRs.
I also suggest the rename of the options - `graph.diff.*` isn't the best name.

**Review by XVilka — APPROVED** [review]

**Review by thestr4ng3r — APPROVED** [review]

**Comment by ITAYC0HEN** [comment]
Don't forget the book :)

**Pull request merged**